### PR TITLE
Minimal expressions API for vortex

### PR DESCRIPTION
**Pull request by jdcasale**
Minimal set of expressions and operators for defining predicates over vortex arrays.

This does **_not_** express recursive expression trees, but instead restricts the algebra to disjunction-of-conjunction (OR-of-AND) of field/value comparison operations.

I tried to restrict this to a manageable set of expressions that we will definitely need as to avoid bloat; the expectation here is that this may grow over time.

Subsequent changes will introduce:

- [ ] A pushdown API for vortex arrays
- [ ] Array pushdown implementations for the various expressions

Supercedes https://github.com/spiraldb/vortex/pull/308

### CODE REVIEW

File: vortex-expr/src/display.rs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+
+            write!(f, "{conj}")?;
+        }
+        Ok(())
+    }
+}
+
+impl Display for ConjunctionExpr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let iter = self.predicates.iter();
```
Comment (gatesn) [inline vortex-expr/src/display.rs]: https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.intersperse

File: vortex-expr/src/expression_fns.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#![allow(dead_code)]
+
+use vortex_dtype::FieldName;
+
+use crate::expressions::Value::Field;
+use crate::expressions::{FieldExpr, PredicateExpr, Value};
+use crate::operators::Operator;
+
+pub fn predicate(left: Value, op: Operator, right: Value) -> PredicateExpr {
```
Comment (gatesn) [inline vortex-expr/src/expression_fns.rs]: I would just get rid of all of these in favour of using `::new(...)`

File: vortex-expr/src/expressions.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
+use serde::{Deserialize, Serialize};
+use vortex_dtype::FieldName;
+use vortex_scalar::Scalar;
+
+use crate::expression_fns::predicate;
+use crate::literal::lit;
+use crate::operators::Operator;
+
+#[derive(Deserialize, Serialize)]
+pub struct DNFExpr {
```
Comment (gatesn) [inline vortex-expr/src/expressions.rs]: Drop `Expr` suffix from everything. It's a bit verbose.

Also DNF -> Disjunction

File: vortex-expr/src/expressions.rs [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    pub conjunctions: Vec<ConjunctionExpr>,
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct ConjunctionExpr {
+    pub predicates: Vec<PredicateExpr>,
```
Comment (gatesn) [inline vortex-expr/src/expressions.rs]: Either have `Conjunction(pub Vec<Predicate>)` or remove the pub here and add accessor field.

File: vortex-expr/src/expressions.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub enum Value {
+    /// A named reference to a qualified field in a dtype.
+    Field(FieldExpr),
+    /// A constant scalar value.
+    Literal(Scalar),
+    /// True if argument is NULL, false otherwise. This expression itself is never NULL.
+    IsNull(FieldExpr),
```
Comment (gatesn) [inline vortex-expr/src/expressions.rs]: What is this for?

File: vortex-expr/src/expressions.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+impl PredicateExpr {
+    pub fn new(left: Value, op: Operator, right: Value) -> Self {
+        Self { left, op, right }
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Deserialize, Serialize)]
+pub struct FieldExpr {
```
Comment (gatesn) [inline vortex-expr/src/expressions.rs]: No point really wrapping this up, e.g. you use Scalar directly

File: vortex-expr/src/literal.rs [hunk 7]
```diff
@@ -0,0 +1,26 @@
+use vortex_scalar::Scalar;
+
+use crate::expressions::Value;
+
+pub trait Literal {
```
Comment (gatesn) [inline vortex-expr/src/literal.rs]: ?

File: vortex-expr/src/scalar.rs [hunk 8]
```diff
@@ -0,0 +1,12 @@
+use vortex_scalar::Scalar;
+
+use crate::expressions::Value;
+use crate::literal::Literal;
+
+pub struct ScalarDisplayWrapper<'a>(pub &'a Scalar);
+
+impl Literal for Scalar {
```
Comment (gatesn) [inline vortex-expr/src/scalar.rs]: ?

File: vortex-expr/src/scalar.rs [hunk 9]
```diff
@@ -0,0 +1,12 @@
+use vortex_scalar::Scalar;
+
+use crate::expressions::Value;
+use crate::literal::Literal;
+
+pub struct ScalarDisplayWrapper<'a>(pub &'a Scalar);
```
Comment (gatesn) [inline vortex-expr/src/scalar.rs]: Why is this here?

File: vortex-expr/src/display.rs [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+}
+
+impl Display for PredicateExpr {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "({} {} {})", self.left, self.op, self.right)
+    }
+}
+
+/// Alternative display for scalars
+impl Display for ScalarDisplayWrapper<'_> {
```
Comment (gatesn) [inline vortex-expr/src/display.rs]: Probably not worth having for now. If we do want it, should add it to the scalar crate maybe called `TypedScalarDisplay`

File: vortex-expr/src/display.rs (hunk 1 above)
Comment (jdcasale) [reply vortex-expr/src/display.rs]: Apparently they're adding this to the standard library soon so using that fn results in a compiler warning -- can do `#![feature(iter_intersperse)]` though

File: vortex-expr/src/expressions.rs [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+)]
+pub struct Conjunction {
+    pub predicates: Vec<Predicate>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Value {
+    /// A named reference to a qualified field in a dtype.
+    Field(FieldName),
```
Comment (gatesn) [inline vortex-expr/src/expressions.rs:30]: As discussed offline, this will need to turn into a FieldPath

**Review by gatesn — APPROVED** [review]
